src/http_handler/handler.cpp

```diff
@@ -0,0 +1,220 @@
+		queryApiParameters.sqlParametersOpt.has_value() &&
+		queryApiParameters.sqlParametersOpt.value().empty() == false;
+
+    if (use_prepared_stmt) {
+    	auto prepared_stmt = con.Prepare(query);
+		if (prepared_stmt->HasError()) {
+			throw HttpHandlerException(500, prepared_stmt->GetError());
+		}
+
+		auto named_values = queryApiParameters.sqlParametersOpt.value();
+
+		auto prepared_stmt_result = prepared_stmt->Execute(named_values);
+		D_ASSERT(prepared_stmt_result->type == QueryResultType::STREAM_RESULT);
+		result = unique_ptr_cast<QueryResult, StreamQueryResult>(std::move(prepared_stmt_result))->Materialize();
+    } else {
+		result = con.Query(query);
+    }
+
+	if (result->HasError()) {
+		throw HttpHandlerException(500, result->GetError());
+	}
+
+	return result;
+}
+
+QueryApiParameters ExtractQueryApiParameters(const duckdb_httplib_openssl::Request& req) {
+	if (req.method == "POST" && req.has_header("Content-Type") && req.get_header_value("Content-Type") == "application/json") {
+		return ExtractQueryApiParametersComplex(req);
+	}
+	else {
+		return QueryApiParameters {
+			ExtractSqlQuerySimple(req),
+			std::nullopt,
+            ExtractOutputFormatSimple(req),
+		};
+	}
+}
+
+std::optional<std::string> ExtractSqlQuerySimple(const duckdb_httplib_openssl::Request& req) {
+    // Check if the query is in the URL parameters
+    if (req.has_param("query")) {
+        return req.get_param_value("query");
+    }
+    else if (req.has_param("q")) {
+        return req.get_param_value("q");
+    }
+
+    // If not in URL, and it's a POST request, check the body
+    else if (req.method == "POST" && !req.body.empty()) {
+        return req.body;
+    }
+
+    return std::nullopt;
+}
+
+OutputFormat ExtractOutputFormatSimple(const duckdb_httplib_openssl::Request& req) {
+    // Check for format in URL parameter or header
+    if (req.has_param("default_format")) {
+        return ParseOutputFormat(req.get_param_value("default_format"));
+    }
+    else if (req.has_header("X-ClickHouse-Format")) {
+        return ParseOutputFormat(req.get_header_value("X-ClickHouse-Format"));
+    }
+    else if (req.has_header("format")) {
+        return ParseOutputFormat(req.get_header_value("format"));
+    }
+    else {
+    	return OutputFormat::Ndjson;
+    }
+}
+
+OutputFormat ParseOutputFormat(const std::string& formatStr) {
+    if (formatStr == "JSONEachRow" || formatStr == "ndjson" || formatStr == "jsonl") {
+    	return OutputFormat::Ndjson;
+    }
+    else if (formatStr == "JSONCompact") {
+    	return OutputFormat::Json;
+    }
+    else {
+    	throw HttpHandlerException(400, "Unknown format");
+    }
+}
+
+QueryApiParameters ExtractQueryApiParametersComplex(const duckdb_httplib_openssl::Request& req) {
+	yyjson_doc *bodyDoc = nullptr;
+
+	try {
+		auto bodyJson = req.body;
+		auto bodyJsonCStr = bodyJson.c_str();
+		bodyDoc = yyjson_read(bodyJsonCStr, strlen(bodyJsonCStr), 0);
+
+		return ExtractQueryApiParametersComplexImpl(bodyDoc);
```

Comment:
bodyDoc is never freed

Reply:
possibly `json_line` as well?

```diff
diff --git a/src/http_handler/handler.cpp b/src/http_handler/handler.cpp
index 88d25a3..e5f7a8b 100644
--- a/src/http_handler/handler.cpp
+++ b/src/http_handler/handler.cpp
@@ -179,6 +179,7 @@ QueryApiParameters ExtractQueryApiParametersComplex(const duckdb_httplib_openssl
         return ExtractQueryApiParametersComplexImpl(bodyDoc);
     }
     catch (const std::exception& exception) {
+        yyjson_doc_free(bodyDoc);
         throw;
     }
 }
@@ -217,6 +218,7 @@ QueryApiParameters ExtractQueryApiParametersComplexImpl(yyjson_doc* bodyDoc) {
 
 static std::string ConvertResultToNDJSON(MaterializedQueryResult &result) {
     std::string ndjson_output;
+    yyjson_mut_doc *doc = nullptr;

     for (idx_t row = 0; row < result.RowCount(); ++row) {
         // Create a new JSON document for each row
@@ -226,6 +228,8 @@ static std::string ConvertResultToNDJSON(MaterializedQueryResult &result) {
         yyjson_mut_doc_set_root(doc, root);
 
         for (idx_t col = 0; col < result.ColumnCount(); ++col) {
+            // Ensure doc is freed on each iteration
+            yyjson_mut_doc_free(doc);
             Value value = result.GetValue(col, row);
             const char* column_name = result.ColumnName(col).c_str();
 
@@ -246,6 +250,7 @@ static std::string ConvertResultToNDJSON(MaterializedQueryResult &result) {
 
         ndjson_output += json_line;
         ndjson_output += "\n";
+        free(json_line);
     }
     return ndjson_output;
 }
```

Reply:
@vhiairrassary shall I apply the patches or you prefer doing so?